cluster/network.go

```diff
@@ -209,3 +231,236 @@ func (c *Cluster) getNetworkPluginManifest(pluginConfig map[string]string) (stri
+		runHost := host
+		errgrp.Go(func() error {
+			return checkPlaneTCPPortsFromHost(ctx, runHost, etcdPortList, c.EtcdHosts, c.SystemImages[AplineImage])
+		})
+	}
+	if err := errgrp.Wait(); err != nil {
+		return err
+	}
+	// Workers need to talk to etcd for calico
+	for _, host := range c.WorkerHosts {
+		runHost := host
+		errgrp.Go(func() error {
+			return checkPlaneTCPPortsFromHost(ctx, runHost, etcdPortList, c.EtcdHosts, c.SystemImages[AplineImage])
+		})
+	}
+	if err := errgrp.Wait(); err != nil {
+		return err
+	}
+	// check controle plane -> Workers
+	log.Infof(ctx, "[network] Running control plane -> etcd port checks")
+	workerPortList := []string{
+		KubeletPort,
+	}
+	for _, host := range c.ControlPlaneHosts {
+		runHost := host
+		errgrp.Go(func() error {
+			return checkPlaneTCPPortsFromHost(ctx, runHost, workerPortList, c.WorkerHosts, c.SystemImages[AplineImage])
+		})
+	}
+	if err := errgrp.Wait(); err != nil {
+		return err
+	}
+	// check workers -> control plane
+	log.Infof(ctx, "[network] Running workers -> control plane port checks")
+	controlPlanePortList := []string{
+		KubeAPIPort,
+	}
+	for _, host := range c.WorkerHosts {
+		runHost := host
+		errgrp.Go(func() error {
+			return checkPlaneTCPPortsFromHost(ctx, runHost, controlPlanePortList, c.ControlPlaneHosts, c.SystemImages[AplineImage])
+		})
+	}
+	return errgrp.Wait()
+}
+
+func checkPlaneTCPPortsFromHost(ctx context.Context, host *hosts.Host, portList []string, planeHosts []*hosts.Host, image string) error {
+	hosts := []string{}
+	for _, host := range planeHosts {
+		hosts = append(hosts, host.Address)
+	}
+	imageCfg := &container.Config{
+		Image: image,
+		Tty:   true,
+		Env: []string{
+			fmt.Sprintf("HOSTS=%s", strings.Join(hosts, " ")),
+			fmt.Sprintf("PORTS=%s", strings.Join(portList, " ")),
+		},
+		Cmd: []string{
+			"sh",
+			"-c",
+			"for host in $HOSTS; do for port in $PORTS ; do nc -z $host $port > /dev/null || echo $host $port ; done; done",
+		},
+	}
+	if err := docker.DoRemoveContainer(ctx, host.DClient, PortCheckContainer, host.Address); err != nil {
+		return err
+	}
+	if err := docker.DoRunContainer(ctx, host.DClient, imageCfg, nil, PortCheckContainer, host.Address, "network"); err != nil {
+		return err
+	}
+	if err := docker.WaitForContainer(ctx, host.DClient, PortCheckContainer); err != nil {
+		return err
+	}
+	logs, err := docker.ReadContainerLogs(ctx, host.DClient, PortCheckContainer)
+	if err != nil {
+		return err
+	}
+	defer logs.Close()
+	if err := docker.RemoveContainer(ctx, host.DClient, host.Address, PortCheckContainer); err != nil {
+		return err
+	}
+	portCheckLogs := getPortCheckLogs(logs)
+	if len(portCheckLogs) > 0 {
+
+		return fmt.Errorf("[netwok] Port check for ports: [%s] failed on host: [%s]", strings.Join(portCheckLogs, ", "), host.Address)
+
+	}
+	return nil
+}
+
+func getPortCheckLogs(reader io.ReadCloser) []string {
+	logLines := bufio.NewScanner(reader)
+	hostPortLines := []string{}
+	for logLines.Scan() {
```

Comment:
shall we care about scanner returning an error?

Reply:
I am not 100% sure, but I added the error check just to be safe